File: drawerbehavior/src/main/java/com/infideap/drawerbehavior/Advance3DDrawerLayout.java
package com.infideap.drawerbehavior;

import android.content.Context;
import android.graphics.Color;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.CardView;
import android.util.AttributeSet;

/**
 * Created by Shiburagi on 21/09/2017.
 */
public class Advance3DDrawerLayout extends AdvanceDrawerLayout {
    private static final String TAG = Advance3DDrawerLayout.class.getSimpleName();

    public Advance3DDrawerLayout(Context context) {
        super(context);
    }


    public Advance3DDrawerLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public Advance3DDrawerLayout(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    @Override
    void updateSlideOffset(CardView child, AdvanceDrawerLayout.Setting setting, float width, float slideOffset, boolean isLeftDrawer) {
        updateSlideOffset(child, (Setting) setting, width, slideOffset, isLeftDrawer);

    }

    void updateSlideOffset(CardView child, Setting setting, float width, float slideOffset, boolean isLeftDrawer) {
        if (setting.degree > 0) {
            float percentage = (setting.degree) / 90f;

            ViewCompat.setX(child, width * slideOffset - (child.getWidth()/2) * percentage * slideOffset);
            ViewCompat.setRotationY(child, (isLeftDrawer ? -1 : 1) * setting.degree * slideOffset);

        } else
            super.updateSlideOffset(child, setting, width, slideOffset, isLeftDrawer);


    }

    @Override
    AdvanceDrawerLayout.Setting createSetting() {
        return new Setting();
    }

    public void setViewRotation(int gravity, float degree) {
        int absGravity = getDrawerViewAbsoluteGravity(gravity);
        Setting setting;
        if (!settings.containsKey(absGravity)) {
            setting = (Setting) createSetting();
            settings.put(absGravity, setting);
        } else
            setting = (Setting) settings.get(absGravity);

        setting.degree = degree > 45 ? 45 : degree;

        setting.scrimColor = Color.TRANSPARENT;
        setting.drawerElevation = 0;
    }

    class Setting extends AdvanceDrawerLayout.Setting {
        float degree;
    }
}

File: drawerbehavior/src/main/java/com/infideap/drawerbehavior/Advance3DDrawerLayout.kt
package com.infideap.drawerbehavior

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import androidx.cardview.widget.CardView

/**
 * Created by Shiburagi on 21/09/2017.
 */
class Advance3DDrawerLayout : AdvanceDrawerLayout {
    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle)

    override fun updateSlideOffset(child: CardView, setting: AdvanceDrawerLayout.Setting?, width: Float, slideOffset: Float, isLeftDrawer: Boolean) {
        updateSlideOffset(child, setting as Setting, width, slideOffset, isLeftDrawer)
    }

    private fun updateSlideOffset(child: CardView, setting: Setting, width: Float, slideOffset: Float, isLeftDrawer: Boolean) {
        if (setting.degree > 0) {
            val percentage = setting.degree / 90f
            child.x = width * slideOffset - child.width / 2.0f * percentage * slideOffset
            child.rotationY = (if (isLeftDrawer) -1 else 1) * setting.degree * slideOffset
        } else super.updateSlideOffset(child, setting, width, slideOffset, isLeftDrawer)
    }

    override fun createSetting(): AdvanceDrawerLayout.Setting {
        return Setting()
    }

    fun setRotation(gravity: Int, degree: Float) {
        val absGravity = getDrawerViewAbsoluteGravity(gravity)
        val setting: Setting?
        if (!settings.containsKey(absGravity)) {
            setting = createSetting() as Setting
            settings[absGravity] = setting
        } else setting = settings[absGravity] as Setting?

        setting?.degree = if (degree > 45) 45f else degree
        setting?.scrimColor = Color.TRANSPARENT
        setting?.drawerElevation = 0f
    }

    inner class Setting : AdvanceDrawerLayout.Setting() {
        var degree = 0f
    }

    companion object {
        private val TAG = Advance3DDrawerLayout::class.java.simpleName
    }
}